### ISSUE TITLE
What should be the minimum Python version to support in swig-4.0?

### ISSUE THREAD

**Issue by wsfulton** [swig/swig#701]
The swig-3.0.x release supposedly supports python 2.0 and is currently documented: 

_SWIG is compatible with most recent Python versions including Python 3.0 and Python 2.6, as well as older versions dating back to Python 2.0. For the best results, consider using Python 2.3 or newer._

We currently only test for python-2.4 to 3.5. 

One of the aims for the next version of SWIG (3.1) is to vastly reduce the number of options supported for Python wrappers. There are just over 50 command line options and they can be viewed by running `swig -python -help`. Many of these options only work for newer versions of python and not all the combinations can be tested. Many of the options require newer versions of Python.

This ticket is an opportunity for SWIG users to express their minimum version required. I would like to start by suggesting that 2.7 should be the minimum version supported as there is still a huge code base that has not moved to python-3 and most 2.x projects can move to 2.7. 2.7 is also a good stepping stone to moving to the very different python-3 world as it has some 3.x features backported into it. If anyone wants support for an older version please say so here and why.

Lastly we don't know when swig-3.1 will be released. It requires volunteers to implement the necessary changes, so if you are able to help, please say so.

**Comment by mromberg** [comment]
Given that python 2.6 reached end of life in 2013:


https://www.python.org/dev/peps/pep-0361/

  It seems unreasonable for swig to have to support it or anything even older.  However 2.7 will be supported until 2020:

http://legacy.python.org/dev/peps/pep-0373/

  So, the suggestion for swig-3.1 to only support python-2.7+ seems very reasonable to me.

P.S.  Also, I may be able to help with swig-3.1.  Let me know what needs doing.

**Comment by mruffalo** [comment]
It might also be worth it to specify a minimum version of Python 3. As per [PEP 392](https://www.python.org/dev/peps/pep-0392/) and [PEP 398](https://www.python.org/dev/peps/pep-0398/), Python 3.2 and 3.3 are now EOL, so supporting 2.7 and 3.4+ seems reasonable.

I seem to remember Red Hat offering 3.3 as part of Red Hat Software Collections (or whatever the name is) for RHEL 6 or 7 or something; is 3.3 the most current Python available there? If so it might be worth it to support 2.7 and 3.3+, since dropping support for 3.3 doesn't really have any benefits. Dropping support for 2.6 and 3.0 - 3.2, though, makes it a bit easier to write code that runs unmodified in both 2.x and 3.x. Off the top of my head, 3.3+ allows the `u` string prefix again, which marks a string as Unicode in 2.x and is a no-op in 3.3+ since "plain" strings are already Unicode code points. (Using the `u` prefix is a `SyntaxError` in 3.0 - 3.2.)

Eventually supporting 2.7 and 3.5+ would be nice also, since 3.5 restores the `%` string formatting operator on the `bytes` type, but it'll be a while before that's feasible.

**Comment by vadz** [comment]
I also have customers using RHEL 6 with its Python 2.6, so it would be nice to continue supporting it for now, especially because I think there is no huge benefit/simplification in dropping support for 2.6, as it's not that different from 2.7.

**Comment by ahnolds** [comment]
I'd vote for 2.7 and 3.3+ range moving forward.  If you keep 2.6 support for RHEL 6, then you might also argue for keeping 2.4 support for RHEL 5 (not end-of-life til March 2017 and extended support til November 2020).  I would argue it makes more sense to go based on the Python release lifespan, and as already mentioned 2.6 is already end of life, [as is 3.2](https://www.python.org/dev/peps/pep-0392/). Besides which, if you don't miss the features of Python 2.7/3.x vs 2.6, you probably won't miss the features of Swig 3.1 vs 3.0.

That said, the following is (I believe) the complete relatively small set of places where Swig currently has code to support 2.6 instead of 2.7 (and 3.3/3.4 instead of 3.5+ as a reference)
- __2.6__
    - [PyCapsules](https://docs.python.org/2.7/c-api/capsule.html) were added in 2.7/3.1 (not in 2.6/3.0), see [here](https://github.com/swig/swig/blob/master/Lib/python/pyhead.swg#L204) for the current swig usage
    - Handling of relative imports using [importlib](https://docs.python.org/2/library/importlib.html), see [here](https://github.com/swig/swig/blob/master/Source/Modules/python.cxx#L822) for the current swig usage
        - Note that there are related checks for the version [here](https://github.com/swig/swig/blob/master/Source/Modules/python.cxx#L890) and [here](https://github.com/swig/swig/blob/master/Source/Modules/python.cxx#L1255)
        - Perhaps this will get cleaner after the namespace package situation from #612 etc. is resolved.
- __3.4__
    - [PyAsyncMethods](https://docs.python.org/3/c-api/typeobj.html#async-object-structures) used [here](https://github.com/swig/swig/blob/master/Source/Modules/python.cxx#L4146)
    - [Matrix multiplication operator](https://www.python.org/dev/peps/pep-0465/), aka the nb_matrix_multiply and nb_inplace_matrix_multiply slots used [here](https://github.com/swig/swig/blob/master/Lib/python/pyrun.swg#L744) and [here](https://github.com/swig/swig/blob/master/Source/Modules/python.cxx#L4209)
- __3.3__
    - [Object finalization](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_finalize), aka the [tp_finalize slot](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_finalize) used [here](https://github.com/swig/swig/blob/master/Lib/python/builtin.swg#L447), [here](https://github.com/swig/swig/blob/master/Lib/python/pyrun.swg#L828), [here](https://github.com/swig/swig/blob/master/Lib/python/pyrun.swg#L1016), and [here](https://github.com/swig/swig/blob/master/Lib/python/pyinit.swg#L188)

**Comment by wsfulton** [comment]
> P.S. Also, I may be able to help with swig-3.1. Let me know what needs doing.

@mromberg, that would be rather good as it'll require a few people to make this happen. I've created a wiki page for 3.1 development at https://github.com/swig/swig/wiki/SWIG-3.1-Development. As described there I'd like to turn on all the optimizations, but this requires getting the ```-O``` flag to work with the test-suite. Travis shows a few failures atm, so a first step would be to get these tests working. All the different features/options for Python needs analysing and decided if they can be turned on by default. That is another area where contributions will be welcome.

**Comment by ojwb** [comment]
> If you keep 2.6 support for RHEL 6, then you might also argue for keeping 2.4 support for RHEL 5 (not end-of-life til March 2017 and extended support til November 2020)

This seems rather a strawman argument - nobody has suggested keeping 2.4 support (not in this ticket, nor anywhere else I've seen).

If someone's sticking with RHEL 5 at this point, they presumably have an ultra-conservative attitude to change, so aren't going to leap onto SWIG 3.1.x or library bindings which can only be generated with SWIG 3.1.x.

And dropping support for < 2.6 means a lot of cruft can go, but the 2.6 vs 2.7 differences don't seem to be major.

One major issue with supporting older versions is whether we can actually test them.  It's pretty much inevitable that support for Python 2.6 will get broken if we don't regularly test with it, and the oldest package I have available (on Debian unstable) is of 2.7 - I suspect that's generally true for current distro versions.  But it looks like there's a PPA used by the travis builds, so at least while that's maintained for the Ubuntu version travis uses we will be able to catch problems promptly.

@wsfulton Are you intending to review minimum version support for other languages?

**Comment by wsfulton** [comment]
I am happy to discuss minimum supported levels for other languages too. Not sure much change is needed though, except for maybe Go and PHP. Probably Javascript too, but we need someone familiar with Javascript to get involved first.

@ianlancetaylor, when we get close to releasing 3.1, do you want to remove support for older Go versions?

**Comment by ianlancetaylor** [comment]
Right now the minimum version of Go supported is 1.2.  This mainly affects the SWIG Makefiles and testing; it doesn't affect the code that SWIG generates.  The Go project only supports Go 1.6 and Go 1.7 right now, but I'm not sure offhand what the distros ship.  The biggest simplification for the current SWIG Makefiles would be to assume Go 1.5.

**Comment by ojwb** [comment]
Given PHP frequently needs security patches and is rarely used outside of a web context, I tend to feel we can stop caring about a version when upstream drops security support for it.  Distros may have packages for a bit longer and attempt to provide security support by themselves, but distros with old PHP should also have an correspondingly old SWIG version.

For Perl we currently say:

> For the best results, it is recommended that SWIG be used with Perl 5.8 or later. We're no longer testing regularly with older versions, but Perl 5.6 seems to mostly work, while older versions don't

I believe that was based on my building an older Perl a while ago and trying it.  I'd be inclined to at least drop 5.6 there, as ISTR there were some issues with it, and we aren't ever testing with anything that old.

**Comment by thorade** [comment]
According to the Python docs, there is some difference between Python 2.6 and 2.7:
https://docs.python.org/3/howto/pyporting.html#drop-support-for-python-2-6-and-older
And I have also seen some projects dropping support for Python before 2.7.
Whoever wants to use very old Python can also use very old SWIG, I would say.

Regarding Python 3, some compatibility stuff was added in 3.3, especially PEP-0414:
https://www.python.org/dev/peps/pep-0414/
So I would suggest to also drop 3.0 and 3.1 and 3.2 at least. 
Maybe even drop 3.3, better put all effort into supporting the latest versions!

But then again, this is just my opinion.

**Comment by ojwb** [comment]
Let's try to keep this focussed on the question that was actually asked:

_"This ticket is an opportunity for SWIG users to express their minimum version required"_

So please just tell us the minimum version (or versions if 2.x and 3.x) which you require and leave it at that.  It's just distracting to have a lot of lobbying for killing off support for older versions from people who don't happen to need them.

@vadz has said he'd find 2.6 support useful, and he's actively involved in helping to maintain SWIG.  I've looked at the differences between handling 2.6 and 2.7 in the SWIG code and there's little if anything - for example, there are **NO** checks for `PY_VERSION_HEX` vs `0x02070000`, but 58 conditional checks for earlier versions.

**Comment by hmaarrfk** [comment]
My vote is 2.7, 3.4

I would also like an option to drop all the `try` statements that add noise to the .py file if I just want to support 3+.

**Comment by ojwb** [comment]
> My vote is [...]

This isn't a vote...

> I would also like an option to drop all the try statements that add noise to the .py file if I just want to support 3+.

That's already possible - see the `moduleimport` attribute added to `%module` in 3.0.11.

**Comment by hmaarrfk** [comment]
I understand it isn't a vote. It is good to keep in mind that numpy is dropping support for 2.7 in Jan of 2019. 

Maybe SWIG should jump the gun and make everything simpler?

Anyway, just a thought.

Thanks for tip for `moduleimport`. Makes things cleaner :D

**Comment by ojwb** [comment]
I'm all for documenting realistic "oldest supported versions" so users can have confidence that what the manual says is supported will actually work.  I think historically we've not been active enough at updating the manual to reflect what we're actually testing with (and if we're not testing something it's going to end up broken sooner or later - the only question is how long until that happens).

But we are testing with Python 2.6 and 2.7 still (we have CI builds, so we'll see right away if a patch breaks them).  While we have CI coverage and there are people who want support for 2.7 or 2.6 and they're willing to do the extra work to keep them around I don't think we should ignore them without very careful consideration.

Dropping Python2 support entirely might work for the numpy developers.  It might even work for numpy users.  But numpy is a very different thing to SWIG, and what works for numpy isn't automatically right for us.

**Comment by hmaarrfk** [comment]
@ojwb I'm happy to help clean up some of the python side of the code.
Looking through it today, it was really dense, with many switches for different versions.

Something that would help is better understanding what `-modern` and `-classic` mean. I've only lived in the python 2.6+ world, and so for me, anything specific to things like `2.4` or `2.5` are totally foreign. I don't even know how I can setup a machine to test those. `-modern` seems to mean >2.2. Classic seems to mean <2.1??? They can also both be set to true?

numpy deciding to move on was an example that a big shift is coming soon, so if the 3.1 release is slated for late 2018, why not just jump the gun by a few month and drop 2.7 all together.
Swig 3.0 will always exist if people want older support. 

They probably don't need all the new C++ features at that point either.

**Comment by hugovk** [comment]
To get a general picture of usage, here's the pip installs from PyPI for last month:

`$ pypinfo --start-date -50 --end-date -23 --percent --pip --markdown "" pyversion`

| python_version | percent | download_count |
| -------------- | ------: | -------------: |
| 2.7            |  73.85% |    362,071,915 |
| 3.6            |  12.92% |     63,337,814 |
| 3.5            |   8.17% |     40,075,882 |
| 3.4            |   4.19% |     20,540,641 |
| 2.6            |   0.71% |      3,480,030 |
| 3.3            |   0.08% |        406,934 |
| 3.7            |   0.07% |        360,437 |
| 3.2            |   0.00% |         23,511 |
| None           |   0.00% |         11,151 |
| 3.8            |   0.00% |          1,766 |

**Comment by hmaarrfk** [comment]
@hugovk That is very informative.

Running an example `python/callback` and looking at `example_wrap.cxx`. I mention the number of times a specific version of python is mentionned in a #define comparison with `PY_VERSION_HEX`
here is a summary of the output

|python_version | count|
|:-------|:-|
| < 2.0 | 2 | 
| >2.0 | 1 | 
| 2.1 | 3 | 
| 2.2 | 17 | 
| 2.3 | 10|
|2.4 | 1| 
|2.5 | 7|
|2.6| 3|
|2.7| 0 |
|3.0| 29 |
|3.1| 0 | 
|3.2| 2 | 
|3.3| 0 |
|3.4| 4 |
|3.5 | 1|
|3.6| 0 |
|3.7|0 |


So in my mind, bv going all the way up to 2.7, we get to simplify the logic by 44 `#define` statements.
By getting rid 2.7 entirely, the logic will be simplified by 73 statements, making it much more understandable what is going on.

**Comment by hmaarrfk** [comment]
Of course, the 2.7, 3 debate wont be solved here, but I can definitely help pruning alot of the older Python versions (<2.5) if people think it is worth it.

**Comment by ojwb** [comment]
So if @vadz still wants 2.6 support and is prepared to put in any work needed to keep it you think we should just ignore him?  That seems unreasonable to me - your grepping shows (as I already said above) that 2.6 vs 2.7 is hardly any difference.

As a tool that's used to make python libraries we should be more conservative about version support than most individual libraries are.  Dropping 2.x entirely at this point is just too aggressive.

There seems to be a clear consensus on dropping support for 2.5 and older (which gets rid of much of the cruft).  I've not seen anyone suggest we should be supporting those here or elsewhere.

The bigger problem is actually that we have far too many options specific to the Python backend, most of them have little to no useful documentation, and the defaults aren't really appropriate for any version of Python actually in use still.  Rationalising this would make far more difference to the code than dropping support for Python 2.6 or even Python 2.x entirely would.  It'd also help with testing - currently there are a huge number of different possible combinations of options, and many of these options potentially interact.

> `-modern` and `-classic`

AIUI, Python had several different iterations on how classes work, which in timeline order you get with `-classic`, no special options, and `-modern` respectively.  At this point `-modern` should just be the default and the other two consigned to history.  There's also `-builtin` which wraps objects as built in types which is faster.  It works with 2.3 and up but works better with 2.5 and up.  Apparently it currently has some incompatibilities, but if we could address those it might be a better default still (and `-modern` dropped entirely too).  I don't know if that's feasible though.

See https://github.com/swig/swig/wiki/SWIG-4.0-Development#simplifying-python-options for some analysis.

**Comment by hmaarrfk** [comment]
Thank you for your detailed reply Olly.

I am not suggesting that SWIg should come to the conclusion of dropping 2.X
but rather I was expressing my personal opinion.

That you for your explanation of classic and modern. I gives me a head
start in understanding the python options.

**Comment by ojwb** [comment]
I've made a start on this in #1261.

Regarding Python 3.x, it looks like nobody's interested in Python 3.2 (or 3.1 or 3.0), but there are a few suggestions above to support >= 3.3, which would allow cleaning up most of the 3.x conditional checks.  Currently we seem to have CI coverage for 3.2 and up.

**Comment by wsfulton** [comment]
It's been 2 years since this question was first asked, I think there is even less of an argument for supporting 2.6 now. I'd like 2.7, 3.3+ only supported. @vadz, I think you were the only one who expressed interest in 2.6 support as 'nice'. Is it really essential to have?

**Comment by vadz** [comment]
No, not essential in the sense that we can just keep using the current version of SWIG, but I think we still have a couple of customers stuck on RHEL 5 with Python 2.6. I wouldn't bother doing anything complicated to keep 2.6 support at this stage, but I believe there are (almost?) no significant differences between it and 2.7 at C API level, so if it doesn't cost anything to keep it, then I would still do it.

**Comment by ojwb** [comment]
At the C API level, there's at least the "capsule" stuff not in 2.6, and 2.6 also needs `(char *)` casts on literal strings in a few more places than 2.7 does.  On the Python side, the import machinery has a whole special version for 2.6.

There's certainly some extra work involved in keeping 2.6 going, and a key issue is it's harder work to do (I can install packages of 2.7 but not of 2.6).  As a practical demonstration, currently the branch I started yesterday passes everywhere except 2.6.

If we're going more aggressive, I suggest we go for >= 2.7 and >= 3.4.  Python 3 early adopters are naturally the types to have moved on quickly to the latest shiny, and similarly to 2.6 I can't install packages of 3.3 to test with locally (but more than that, there aren't Python 3.3 packages in **ANY** Debian or Ubuntu release as far as I can see).  Also 3.3 [reached upstream EOL in September 2017](https://www.python.org/dev/peps/pep-0398/#x-end-of-life).

**Comment by ojwb** [comment]
On the subject reviewing the decision here in the light of passing time, I'm going to propose we drop PHP5 for 4.0.0.

PHP 5.6 is security supported by PHP upstream until the end of this year (http://php.net/supported-versions.php).  Even if we get 4.0.0 out before the end of this year, there's not going to be much time between its release and PHP5 being out of security support, and it doesn't make much sense to carry a dead PHP major version for almost a whole SWIG release cycle.

We have PHP7 support now, but because the C API changes were so extensive this is a fork of the PHP5 backend, so it'll be easy to drop PHP5 without risk of affecting PHP7.

**Comment by vadz** [comment]
> At the C API level, there's at least the "capsule" stuff not in 2.6

Ah, yes, I totally forgot about this, but you're right and it's pretty important as without capsules it's impossible (AFAIK) to ensure that only a single copy of common types is used in multi-module libraries, so it's a serious limitation.

So you've convinced me, let's drop support for it, we'll deal with any customers stuck with it in some way, but it's not a reason to not drop support for it in SWIG 4.0.

Also, I definitely agree that 3.4 is low enough for Python3 version requirement, nobody uses earlier versions any more by now.

**Comment by ojwb** [comment]
I'm wondering if `Lib/python/pybackward.swg` can be removed for 4.0 - it's apparently "adding backward compatibility macros" and I notice it contains some suspect code:

```
  Type v;
  int res = SWIG_AsVal(Type)(obj, &v);
  if (!SWIG_IsOK(res)) {
    /*
      this is needed to make valgrind/purify happier. 
     */
    memset((void*)&v, 0, sizeof(Type));
```

So in order to make valgrind and purify happier we do a byte-level zero of a variable of type `Type`, which appears to be an arbitrary type - that's not OK for arbitrary classes and modern GCC and clang will both warn about this (which leads me to think that this compatibility code isn't widely used since nobody has reported such warnings here and we've had multiple reports for other cases).

This file was added in f39d358984d2d4f3c015b46d3bc2df0a7ac96a9f back in Jan 2006 (so over 12 years ago) and hasn't been touched since.  The commit message was "add more backward comp macros/fragments/methods".  The only useful information I could find was in the follow-up commit 3de733ebc449ec06853d2fc097952ab28390cf2b.

So this is for compatibility with what are now really ancient versions of SWIG (definitely < 1.3.28) and 4.0 seems a good time to clean up the codebase by removing such things.

Thoughts?

**Comment by vadz** [comment]
FWIW there are apparently some references to this fragment in [Subversion](https://svn.apache.org/repos/asf/subversion/trunk/subversion/bindings/swig/include/svn_global.swg). This is the only occurrence I could find using [Debian](https://codesearch.debian.net/search?q=SWIG_As_frag) or [GitHub](https://github.com/search?p=35&q=SWIG_As_frag&type=Code) search, but even if there are no others, breaking Subversion bindings without any good reason doesn't look like a good idea...

**Comment by mromberg** [comment]
Major version number increases do seem like a good spot to break backwards compatibility.   But perhaps it could be done by first deprecating the whole file with a pre-processor #warning and then removing it after a bit of time?

**Comment by ojwb** [comment]
We could just tell the Subversion people that they're using obsolete interfaces which are only provided for compatibility with 12+ year old versions of SWIG.

Skimming their interface file, I suspect a lot of the stuff they're doing is working around long-fixed SWIG issues and deficiencies.

I'm not sure we can just add a `#warning` as this file is currently always included for Python.  We could stop doing that but leave it there so people using it would have to add an explicit `%include <pybackward.swg>` or similar, but that's seems a terrible compromise - affected users still have to make changes to their interface files, and we still have to keep this cruft around for even longer.  

If we're going to do something that requires changes to (a very small number of) user interface files, I think it's better to make them migrate off these obsolete interfaces entirely.

There's no test coverage at all for this code (I tried replacing `Lib/python/pybackward.swg` with an empty file and both `make check-python-examples` and `make check-python-test-suite` pass), and adding proper test coverage would likely be more work than migrating users off this.

**Comment by wsfulton** [comment]
Let's remove the pybackward.swg file and help users migrate off. Providing a patch to Subversion would be the most helpful. The simplest and easiest would be to add pybackward.swg into their repo, but I guess we can probably do better than that.

**Comment by ojwb** [comment]
I fixed the travis lua macOS build earlier (85c094a325338de815fc623892998eef28742ee7) and in doing so noticed that we claim to support Lua 5.0 and up but only have CI builds for 5.2 and 5.3.  I tried hard to keep compatible with 5.0 and 5.1 in my fix, but I can't easily test with either.

There are various special-cases in the code for Lua 5.0 and 5.1, so officially aligning support with what we can actually test would allow some clean-up, and looks very sensible based on end-of-life dates:

https://www.lua.org/versions.html

> Lua 5.2 [...] The last release was Lua 5.2.4, released on 07 Mar 2015. There will be no further releases of Lua 5.2.
>
> Lua 5.1 [...] The last release was Lua 5.1.5, released on 17 Feb 2012. There will be no further releases of Lua 5.1.
>
> Lua 5.0 [...] The last release was Lua 5.0.3, released on 26 Jun 2006. There will be no further releases of Lua 5.0.

So we're close to 12 years past the final 5.0.x and more than 6 years past the final 5.1.x.  5.2.x is end-of-life as well, but doesn't seem to require any special handling and is at least currently easy to test with in travis.

To be clear, I'm proposing we aim to support Lua >= 5.2.0 (though I don't have a strong argument for keeping 5.2.x support if those more involved in Lua backend work think we should drop that too).

**Comment by wsfulton** [comment]
I've got to object to some of these 3.x versions that we are dropping support for which are not that old. What is the technical reason for not supporting 3.3? I don't think there are any. The more older versions we can support the better for users of SWIG. We should balance this up with the burden of maintaining them by SWIG developers. Very few readers will be reading this discussion so we should provide as many older versions as we can given this balance. Given the Python 3 API stabilised in 3.2 (see PEP3003), we could keep 3.2 and 3.3 support without any real burden that I can think of.

Likewise for older Lua; I don't know of any technical reason to drop support, so just leave it in.

**Comment by hugovk** [comment]
I've noticed for many projects there are very few Python 3.3 users, if any.

It looks like people who are already on the Python 3.x track are much more likely to upgrade.

It's true that the technical burden for supporting 3.3 is much lower than, say, 2.6 or 3.2, and that there are fewer benefits from upgrading.

However, by only supporting the non-EOL versions, then you can make use of newer features of Python. This will be especially true when it's [time](https://pythonclock.org/) to [drop 2.x](http://python3statement.org/) entirely.

And one concrete technical benefit from dropping unsupported versions is it reduces the time the CI takes to build. For example, from 14.5 hours to 13.25 hours. That's already a really long build, so everything helps.

**Comment by vadz** [comment]
@wsfulton I just don't think anybody uses Python 3 < 3.4. There, AFAIK, no legacy Linux systems shipping with these versions and people who use Python 3 tend to use more up-to-date versions anyhow -- those who prefer stability are still using Python 2.x.

**Comment by wsfulton** [comment]
SWIG isn't here to unnecessarily force users to migrate from older lua/python versions, we are here to aid and assist users wishing to use these languages from C++. It is quite normal in my world, anyway, unfortunately, for folk to be using software and operating systems that are over 15 years old. If there is just one person using 3.2/3.3  (and there is bound to be), there is no convincing reason I've heard so far to drop support as supporting these versions really is not a burden.

**Comment by ojwb** [comment]
* We've just had a two year long (to the day!) discussion involving a significant number of people in which nobody expressed any desire for Python 3 < 3.4.
* Then 3 weeks ago I explicitly proposed Python >= 3.4 for SWIG 4.0.0 in this very issue, and nobody spoke up against that as a plan (and you yourself replied to this issue in response to a newer comment just over a week ago, so it seemed reasonable to think you agreed with this plan).
* Python 3.3 is no longer supported by Python upstream.
* Python 3.3 no longer seems to be shipped by any distro (e.g. it's not in Debian oldstable even).
* Python 3.3 is at least still available in travis-ci for now, but it seems unlikely to remain so for much of the life of SWIG 4.0.x.
* The sheer number of CI build variants we have for Python has become completely ridiculous.  Good CI coverage is great, but each additional CI build means you have to wait that bit longer to know if a PR is good to merge.  If I push a bad commit, chances are I'll be asleep by the time I know so master ends up broken for much longer.  And the general low-level flakiness of travis becomes more of an issue - I seem to have to force retry of at least one build a lot of the time.
* We don't have enough developer time to do everything - the number of open issues keeps climbing; worse, the number of open PRs keeps climbing (though more slowly) - sadly it seems we don't even have the resources to keep on top of people who have gone to the trouble of providing a patch.

I'm all for supporting older versions which there's actual user interest in, but keeping support for versions "just in case" does have a real burden.  And I really believe this is a problem, in particular because it raises the bar to casual contribution, and most long-term contributors to projects start off as casual contributors.  More long-term contributors could make a very big difference.

The additional burden is often small for each individual case, but in total we end up with a code base riddled with conditionals which makes the code much harder to follow (I struggle to understand the Python backend code and I've been involved in SWIG development for well over a decade now - I shudder to think how that looks to somebody new to the codebase).  Death by a thousand paper cuts.

Developers are also expected to support old versions of languages that they can't simply install locally.  I find myself reluctant to touch anything involving some ancient version of a language - if it's also too ancient to have CI coverage (e.g. Lua 5.0, Lua 5.1) then there's no way to know if I broke it anyway; if there's CI coverage then at least I know breakage should be caught, but because the CI takes so long due to the sheer number of variants I know from experience it'll be painful.  And so stuff doesn't get cleaned up - an example is my recent configure change to fix for the Lua macOS CI build - I'd estimate that took about twice as long to do due to having to carefully preserve the special-case handling of Lua 5.0 and 5.1, and even then I don't know for sure that they still work.  @vadz asked why I didn't simplify things by switching to pkg-config for Lua - the answer was that I have no idea if that works for Lua 5.0 and 5.1, and no easy way to find out.

And on the other side, I don't believe anyone will seriously expect SWIG 4.0.0 to support Lua 5.0 - if you're conservative enough to stick with Lua 5.0 at this point, you aren't going to leap to a new SWIG major version.  So that's extra work I'm doing which benefits nobody, which is demoralising.  Maybe you disagree and think there are some such users, but even then you'd have to agree that there are going to be very, very few, and that they could just stick with SWIG <= 3.0.12.

The quality of SWIG has improved significantly since we got the testsuites passing 100% for almost all languages, and then set up CI to make sure that stays the case.  CI is now a requirement for a new backend, and we've also said it's a requirement for supporting newer node.js versions.  I think going forward it should also be a requirement for supporting **older** versions of a language as well.  We can perhaps consider exceptions if there's a sufficiently good reason (we'll have to make exceptions for the tail end of a release series anyway in some cases, since we can't control when travis moves to a newer base OS version and we lose language versions which we were getting from the old OS version but aren't in the new OS version).  However the burden of proof needs to be on those making a case for keeping a version despite having no CI.

**Comment by ojwb** [comment]
So what exactly was the point of starting a two-year long discussion with

> This ticket is an opportunity for SWIG users to express their minimum version required.

only to wrap it up with

> Very few readers will be reading this discussion so we should provide as many older versions as we can

?

**Comment by wsfulton** [comment]
Supporting Python 3.2/3.3 *really* is not a burden in code complexity, nor does it limit what we can do going forwards and it requires a paltry 3% extra Travis compute time. I'll repeat my logic: if older versions aren't a burden on us, then keep the support (it's already there!), it will only help users and that is the number one priority. If you think older Lua versions are truly a burden given there is no CI, then fine, drop them, but I can't agree on Python 3.2/3.3.

> So what exactly was the point of starting a two-year long discussion with
> 
>>     This ticket is an opportunity for SWIG users to express their minimum version required.
The conclusion is Python 2.7 is the minimum version required and we're getting unnecessarily distracted by later versions.

> 
> only to wrap it up with
> 
>>     Very few readers will be reading this discussion so we should provide as many older versions as we can
Do not selectively quote my sentence. You forgot a critical end part of the sentence: _given this balance._. Hopefully my re-iteration around burdensome work should make my thoughts clear. 

If you think that a version is not supported upstream and that is the only criteria to choose what will keep users happy, you are ill-informed as unsupported versions are used all the time.

**Comment by hugovk** [comment]
Here's the pip installs from PyPI for May 2018:

| python_version | percent | download_count |
| -------------- | ------: | -------------: |
| 2.7            |  68.85% |    328,874,156 |
| 3.6            |  19.54% |     93,322,182 |
| 3.5            |   7.60% |     36,303,604 |
| 3.4            |   3.28% |     15,684,843 |
| 2.6            |   0.58% |      2,752,553 |
| 3.7            |   0.10% |        482,967 |
| 3.3            |   0.05% |        217,239 |
| 3.2            |   0.00% |         19,482 |
| None           |   0.00% |          9,460 |
| 3.8            |   0.00% |          3,337 |
| Total          |         |    477,669,823 |

Source: `pypinfo --start-date 2018-05-01 --end-date 2018-05-31 --percent --markdown "" pyversion`

Twice as many downloads for the as-yet-unreleased Python 3.7 than for all of EOL 3.2 and 3.3.

Supporting 3.2 is burdensome due to its odd Unicode handling. Python 3.3 not so much.

Debian 7 had 3.2 but went EOL 2018-05-31. The others have 3.4 or newer.

**Comment by mromberg** [comment]
> I'm all for supporting older versions which there's actual user interest in, but keeping support for versions "just in case" does have a real burden. And I really believe this is a problem, in particular because it raises the bar to casual contribution, and most long-term contributors to projects start off as casual contributors. More long-term contributors could make a very big difference.

> CI is now a requirement for a new backend, and we've also said it's a requirement for supporting newer node.js versions. I think going forward it should also be a requirement for supporting older versions of a language as well.

  As a casual swig contributor, I tend to agree with these specific points and your overall argument in general.

  I imagine I'm fairly typical swig contributor in that I usually have some issue with swig that affects a project I am working on.   I have usually diagnosed the problem and can fix it for my specific case which will involve a certain C/C++ compiler and version of usually one of the swig supported backends.  I can also set up swig and figure out how to make the test suite work using this configuration.

  But, there is no way I would ever locally set up a complete test suite for swig.   That level of effort is beyond what I as a casual contributor would be willing to do.   I use to think that more regular swig developers had done this.  But, it seems that this is also not the case.   So, the only practical way to know if your change breaks something is to run the full CI test suite via github.

  When one is working on a swig feature that relates to say python then a swig developer does need to see if ones changes break anything related to swig's supported python versions.   And yes I fully agree that if the CI does not test the version then it is really not supported anyway.  Without the CI tests it is really difficult to tell that your new changes use some language feature that is not universal across all supported swig configurations.   So the CI pretty much has to be used as part of the process of developing a new patch.   It is not just a final sanity check on the back end.

  And the swig CI suite takes a long time to run.   A very, very long time to run.   It would be nice if there was some easy way to only run a subset of it while developing a pull request.   Because most of the time, the chance of a change in say the ruby backend will never affect any of the other languages.  Yet the only way to test all the supported ruby configurations it to run and wait for the entire CI suite to complete.   I understand that it may not be practical to do anything like only running CI on the ruby tests unless one has set up CI locally.   And a casual contributor is probably never going to do that.

  In summation, I agree with the sentiment that all *supported* swig configurations should have CI tests.   If there are not CI tests then the configuration is not really supported anyway.   There is no way any swig developer casual or otherwise can tell if changes work to any degree without these tests.  And the longer the tests take to run the more onerous it is for a casual contributor to prepare a pull request.   So, anything that could be done at all to reduce the time it takes for CI to run would be a good thing.  One obvious way to reduce this time is simply reduce the number of supported configurations.   But another workable scheme would be to easily allow for only a subset to be run while developing a patch to swig.

**Comment by ojwb** [comment]
> Do not selectively quote my sentence. You forgot a critical end part of the sentence: given this balance.. Hopefully my re-iteration around burdensome work should make my thoughts clear.

I wasn't intentionally doing that.  I really am wondering what on earth this whole long-lived discussion was for.  It seems that you asked for user feedback, but then the conclusion is that we have to support every possible version because we can't know for sure that there isn't somebody somewhere who might be using it.

> If you think that a version is not supported upstream and that is the only criteria to choose what will keep users happy, you are ill-informed as unsupported versions are used all the time.

Now you are the one selectively quoting!  That was one of seven (7) bullet points I made in my comment above.  Obviously it isn't the only criterion, but it's a valid point to keep in mind.

We have a very real problem with opportunity cost here - we really don't have the resources to do everything we'd like to and so we need to prioritise.  If you choose not to decide, you still have made a choice.

And because we're choosing (implicitly or explicitly) to prioritise trying to support obsolete versions of languages which at best a very small minority of users and potential users care about, we're sucking up time and energy and enthusiasm which could be usefully directly at dealing with issues which affect large numbers of users and potential users.

For example, we only support a really ancient version of node.js which very few people use, but the patch for supporting newer versions is blocked on CI coverage.  To a significant extent that's not helped by node.js people who seem to have a lot of time and energy for arguing on the PR why we should make a special exception for them and apply the patch without CI coverage, but much less for helping hook up CI coverage.  But I've spent several hours lately just cleaning up support for Python 2.x versions much of which should have been removed years ago - if I'd not been doing that I could probably have found somebody with node.js knowledge to virtually sit down with and actually sort that out.

Anyway I don't think I'm going to persuade you, and you have far more of a claim on the dictator-for-life hat than anyone else still around, so feel free to revert whatever you want of my work and redo it however you like.

I'm going to stop trying to help out with SWIG 4.0.0.  Every time I try to push things along where it seems there's a clear consensus on what to do it seems you tell me to stop.  I'm going to refocus my energy on projects where I might actually get some enjoyment and a sense that what I'm doing is actually appreciated rather than just annoying.

**Comment by vadz** [comment]
@wsfulton There is an argument to be made about supporting old versions, of course, but you seem to totally neglect the other side of the argument and believe there is no harm in keeping support for the old versions indefinitely. This couldn't be further from truth. Python backend code has become completely unmaintainable partly as the result of this (although, to be fair, for most part because of the ridiculous number of options). I'm pretty sure that this discourages contributors, if only because it definitely discouraged me from working on it and I was really glad that @ojwb took the challenge to modernize and clean it up and it would be extremely disappointing if his efforts were wasted. I don't care about Python 3.2 and 3.3 one way or the other, but for me there is no question that it's up to Olly to decide what to do about them because he is the one spending time and energy actually working on this code.

Please consider also that people using 15 year old versions of Python and Lua are also more than likely to use 10 year old version of SWIG which, of course, won't magically stop supporting the old versions just because SWIG 4.0 doesn't support them any more. And you can be sure that these people won't upgrade to SWIG 4.0 any time soon.

Of course, reasonable people may disagree about how to exactly balance the competing goals of supporting as many of different languages versions as possible and making SWIG development/testing as straightforward as possible, but IMHO you're putting the cursor way too far towards the former -- and I'm saying this as a primary maintainer of a project notorious for being very conservative in its support of old systems/compilers.

**Comment by wsfulton** [comment]
@mromberg, it is actually quite easy to set up linux machines for development, and the main SWIG developers do (or should!), especially anyone modifying code outside of a target language module. I've put my notes at https://github.com/swig/swig/wiki/Getting-Started for Ubuntu. You can have the environment ready for the 14 most used languages using just one command line! The different versions of Python and everything else that Travis runs is dead easy to install with this OS too as you just run the appropriate install commands from https://github.com/swig/swig/blob/master/Tools/travis-linux-install.sh. 

But let's please discuss testing/developing/CI in general in another ticket or on the swig-devel mailing list, this issue is beginning to get rather long and off topic.

**Comment by wsfulton** [comment]
> I wasn't intentionally doing that...

@ojwb I do apologise for misreading you. It looked like you had deliberately partially quoted me as the partial quote conveys quite a different meaning and when combined with the tone of your post seemed aggressive.

@hugovk, thanks for the stats showing there are a few tens/hundreds of thousand Python users still using earlier EOL 3.2/3.3. No doubt a much smaller number will be relying on SWIG in that list, but it's useful to get some idea of numbers/percentages. It confirms one thing I think everyone will agree on and that is there are a minority of users who do use EOL software. I don't envisage everyone agreeing on how important this minority of users is. Backwards compatibility is an area of tension and disagreement in many projects, I hope that we all realise that and agree to disagree.

While I am conservative in the versions I'd like to support as @ojwb mentions, I'd like to think that it isn't particularly conservative. We're supporting versions released approx 10-15 years ago. SWIG and gcc are both compilers and if you think about it gcc still supports K&R C from 40 years ago. We're nothing like that conservative! And quite right too given the necessity and smaller number of contributors.  Also, consider the popular Python deadsnakes [PPA](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa) used by the Ubuntu community (and SWIG for our Travis testing).  Remarkably, it contains a freshly built Python-2.3 (released 2003) for the new Ubuntu 18.04 released a few weeks ago which goes EOL in 2023. So clearly there are folk who think this old version of Python is important in a cutting edge environment.

> ... but much less for helping hook up CI coverage. But I've spent several hours lately just cleaning up support for Python 2.x versions much of which should have been removed years ago <snip>

> ... we're sucking up time and energy and enthusiasm which could be usefully directly at dealing with issues which affect large numbers of users and potential users.

@ojwb - very much in agreement with both of these. We all want to remove the older 2.x versions holding up development, but I don't agree that supporting 3.2/3.3 is holding us back. There isn't a single line of code in SWIG that is 3.2 or 3.3 specific, so I have restored support for these two versions.

> I'm going to stop trying to help out with SWIG 4.0.0 ... <snip> ...  a sense that what I'm doing is actually appreciated rather than just annoying.

@ojwb, on the contrary, I was actually very happy and appreciatve that someone else finally brought some momentum around implementing the 4.0 changes discussed over the last few years.  It doesn't mean I agree with some of the supported versions removed. I can see that it would be annnoying for you having me challenge the versions to be supported a couple of weeks after you committed the code, so I apologise for that. I'm just frustrated with myself as I didn't follow this issue, nor many others, very closely over the last few weeks, because I was 100% focused on the Doxygen merge as I know it has not had the due attention on it from me in the past.  Also, I'm certainly am not annoyed by your work.  You contribute a lot of change and given this large volume you can only expect in a healthy development environment that there is going to be some small amount of disagreement and back and forth on some of it.  I'd appreciate it if you returned to 4.0.0 dev, but if you want to find some other less contentious area to work on that would be equally valuable.

**Comment by wsfulton** [comment]
> There is an argument to be made about supporting old versions, of course, but you seem to totally neglect the other side of the argument and believe there is no harm in keeping support for the old versions indefinitely.  This couldn't be further from truth. Python backend code has become completely unmaintainable partly as the result of this (although, to be fair, for most part because of the ridiculous number of options). ... 

@vadz, I hope I don't come across like that and I find this statement completely puzzling.  Please read the very beginning of this issue where you'll see I created this issue pushing for minimum support to be python-2.7 for the very unmaintainable reasons you quote. Python-2.7 is a 9 year old release - that's far from indefinite and newer than the 10 year old python-2.6 you initially went for.

> Please consider also that people using 15 year old versions of Python and Lua are also more than likely to use 10 year old version of SWIG which, of course, won't magically stop supporting the old versions just because SWIG 4.0 doesn't support them any more. 

@vadz Hah! No, not the clients I deal with, they want their cake and eat it! They want their cutting edge C++17 to be wrapped for their users who are using ancient versions of target languages and cannot have backward compatibility breakage. Perhaps this is a difference in attitude wrt some closed source vs open source users? SWIG supports both types of users.  Also see the deadsnakes comment earlier.

> Of course, reasonable people may disagree about how to exactly balance the competing goals of supporting as many of different languages versions as possible and making SWIG development/testing as straightforward as possible, 

Agreed it is difficult getting the balance right, but making an informed decision is at the heart of my (moderately) conservative approach. 

> but IMHO you're putting the cursor way too far towards the former

@vadz In summary and at risk of sounding like a broken record, it depends:
- Yes indeed as a default when there is very little downside to supporting
  many versions, eg supporting the rather old JDK 5 where the C API simply
has not changed and the improvements in the Java language are insignificant/cosmetic.
- No when supporting older versions is inhibiting maintenance and development (like Python <2.7).

**Comment by wsfulton** [comment]
Closing with 2.7 as the minimum version to be supported. 3.0 and 3.1 from the 3.x series will also be dropped. Travis testing no longer tests the unsupported versions.

**Comment by vadz** [comment]
Just for the record, and for potential benefit of others who might end up in the same situation and find this issue while searching for the solution: I finally had to restore support for Python 2.6 for one of my projects as it still has to support RHEL 6 with its system Python version, so here is [my branch](https://github.com/vadz/swig/tree/python-2.6) where I did it. I'm not completely sure if I didn't revert too much, nor that I didn't break something, but the SWIG unit test suite still passes with both 2.7 and 2.6 and the projects own tests pass too.
